docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,861 @@
+    secretBindingRef:
+      name: core-aws
+    providerConfig:
+      apiVersion: aws.cloud.gardener.cloud/v1alpha1
+      kind: InfrastructureConfig
+      networks:
+        vpc: # specify either 'id' or 'cidr'
+        # id: vpc-123456
+          cidr: 10.250.0.0/16
+        internal:
+        - 10.250.112.0/22
+        public:
+        - 10.250.96.0/22
+        workers:
+        - 10.250.0.0/19
+      zones:
+      - eu-west-1a
+    workerPools:
+    - name: pool-01
+      autoScalerMin: 2
+      autoScalerMax: 2
+      providerConfig:
+        apiVersion: aws.cloud.gardener.cloud/v1alpha1
+        kind: WorkerPoolConfig
+        machineImage:
+          name: CoreOS
+          ami: ami-d0dcef3b
+        machineType: m4.large
+        volumeType: gp2
+        volumeSize: 20Gi
+        zones:
+        - eu-west-1a
+  kubernetes:
+    version: 1.11.0
+    ...
+  dns:
+    provider: aws-route53
+    domain: johndoe-aws.garden-dev.example.com
+  maintenance:
+    timeWindow:
+      begin: 220000+0100
+      end: 230000+0100
+    autoUpdate:
+      kubernetesVersion: true
+  backup:
+    schedule: "*/5 * * * *"
+    maximum: 7
+  addons:
+    kube2iam:
+      enabled: false
+    kubernetes-dashboard:
+      enabled: true
+    cluster-autoscaler:
+      enabled: true
+    nginx-ingress:
+      enabled: true
+      loadBalancerSourceRanges: []
+    kube-lego:
+      enabled: true
+      email: john.doe@example.com
+    monocular:
+      enabled: false
+```
+
+:information: The specifications for the other cloud providers Gardener already has an implementation for looks similar.
+
+### CRD definitions and workflow adaptation
+
+In the following we are outlining the CRD definitions which define the API between Gardener and the dedicated controllers.
+After that we will take a look at the current [reconciliation](https://github.com/gardener/gardener/blob/master/pkg/controller/shoot/shoot_control_reconcile.go)/[deletion](https://github.com/gardener/gardener/blob/master/pkg/controller/shoot/shoot_control_delete.go) flow and describe how it would look like in case we would implement this proposal.
+
+#### Custom resource definitions
+
+Every CRD has a `.spec.type` field containing the respective instance of the dimension the CRD represents, e.g. the cloud provider, the DNS provider or the operation system name.
+Moreover, the `.status` field must contain
+
+* `observedGeneration` (`int64`), a field indicating on which generation the controller last worked on.
+* `state` (`*runtime.RawExtension`), a field which is not interpreted by Gardener but persisted; it should be treated opaque and only be used by the respective CRD-specific controller (it can store anything it needs to re-construct its own state).
+* `lastError` (`object`), a field which is optional and only present if the last operation ended with an error state.
+* `lastOperation` (`object`), a field which always exists and which indicates what the last operation of the controller was.
```

Comment:
(This is a copy of an eventually resolved comment)

I didn't see any [event](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#event-v1-core) usage on the gardener, I'm curious to hear if there is a historical reason, and a decision behind not using events.

But if not so, why not using events instead of having these `last*` fields. I think the right thing to do here is to use events since these are actual events and not statuses.

Reply:
Hm, strange, I'm sure I have replied to that, but it seems to be not visible any more. Anyway: Gardener already uses `Event`s, you can see them by doing `kubectl describe shoot`. For every operation it writes a "start" and an "error"/"success" event. It also writes events when a Shoot is maintained.

The `last*` fields cannot be dropped as they will be needed for communication between Gardener and the extension controllers. Parsing/finding events is hard and not very efficient.